Multithreading Render Passes Sample

PavelRudko-ARM opened the pull request:
## Description

This sample shows how we can benifit by recording different render passes in multiple threads.

Please see [tutorial](https://github.com/KhronosGroup/Vulkan-Samples/blob/4f3a7b3c8b7f31dfd1479d66fca908d7f802aff4/samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md).

This PR adds the following framework changes:

-  Orthographic camera (in this case to render the shadow map)
- `CommandBuffer` now allows to specify render pass and framebuffer in `begin` and `begin_renderpass`
- `RenderContext` allows to submit multiple command buffers at once
- 'GeometrySubpass' provides more control over resources binding
- Added a helper for setting viewport and scissor to reuse in in multiple samples

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build and run on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules

Tested on:

- Huawei P20 (Android)
- Desktop (Windows)

PavelRudko-ARM committed:
Multithreading Render Passes

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/Vulkan-Samples?pullRequest=138) <br/>All committers have signed the CLA.

JoseEmilio-ARM commented:
Please add the sample to the list in https://github.com/KhronosGroup/Vulkan-Samples/blob/master/samples/CMakeLists.txt#L77 (just above AFBC)

PavelRudko-ARM committed:
Added this sample to the sample list

gary-sweet reviewed (changes requested):
I'm seeing a bunch of repeated validation layer errors with this change:

```
[error] [framework/core/instance.cpp:41] 0 - UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object: 0x3df73d00 (Type = 6) | Submitted command buffer expects VkImage 0x170000000017[] (subresource: aspectMask 0x2 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_UNDEFINED.

[error] [framework/core/instance.cpp:41] 0 - VUID-VkPresentInfoKHR-pImageIndices-01296: [ VUID-VkPresentInfoKHR-pImageIndices-01296 ] Object: 0x3d3a36f0 (Type = 4) | Images passed to present must be in layout VK_IMAGE_LAYOUT_PRESENT_SRC_KHR or VK_IMAGE_LAYOUT_SHARED_PRESENT_KHR but is in VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL. The Vulkan spec states: Each element of pImageIndices must be the index of a presentable image acquired from the swapchain specified by the corresponding element of the pSwapchains array, and the presented image subresource must be in the VK_IMAGE_LAYOUT_PRESENT_SRC_KHR layout at the time the operation is executed on a VkDevice (https://github.com/KhronosGroup/Vulkan-Docs/search?q=VUID-VkPresentInfoKHR-pImageIndices-01296)

[error] [framework/core/instance.cpp:41] 0 - VUID-VkImageMemoryBarrier-oldLayout-01197: [ VUID-VkImageMemoryBarrier-oldLayout-01197 ] Object: 0x3de0dda0 (Type = 6) | For VkImage 0x190000000019[] you cannot transition the layout of aspect=2 level=0 layer=0 from VK_IMAGE_LAYOUT_DEPTH_ATTACHMENT_OPTIMAL when the previous known layout is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL. The Vulkan spec states: oldLayout must be VK_IMAGE_LAYOUT_UNDEFINED or the current layout of the image subresources affected by the barrier (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkImageMemoryBarrier-oldLayout-01197)
```

PavelRudko-ARM committed:
Fixed layout transition issues

gary-sweet commented on samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md:
```diff
@@ -0,0 +1,103 @@
+
+One way to use multi-threading with multiple render passes is to create a separate primary level command buffer for each of them. In this case command buffers can be recorded independently and then submitted to the queue all at once using ``vkQueueSubmit``.
+
+A better approach is to use secondary level command buffers, which helps to reduce the overhead and get even better performance. First, both of the passes are recorded into two separate secondary command buffers using two threads. Then, we can just reference them in the primary command buffer via ``vkCmdExecuteCommands``.
```
This is an Arm specific statement. For us, the primary buffer mode is slightly faster than secondary.

PavelRudko-ARM replied:
I have rephrased it.

gary-sweet commented on samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md:
```diff
@@ -0,0 +1,103 @@
+
+**Avoid**
+
+* Submitting multiple primary command buffers to the queue at once, when secondary command buffers can be used.
```
Arm specific again

PavelRudko-ARM replied:
I have removed it. However, it leaves us with no "Avoid" section. I don't know how bad it is, probably I should think of something else for a substitution.

gary-sweet commented on samples/performance/multithreading_render_passes/multithreading_render_passes.cpp:
```diff
@@ -52,7 +52,7 @@ bool MultithreadingRenderPasses::prepare(vkb::Platform &platform)
 	}
 
-	load_scene("scenes/bonza/Bonza4X.gltf");
+	load_scene("scenes/sponza/Sponza01.gltf");
```
Did you intend to change the scene here?
This scene is a lot lighter and I don't really see any performance difference between the modes now.

PavelRudko-ARM replied:
No, sorry. It's fixed now.

gary-sweet commented:
The validation errors appear to be fixed now. Thanks

PavelRudko-ARM committed:
Reverted unintentional scene change

JoseEmilio-ARM commented on samples/performance/multithreading_render_passes/multithreading_render_passes.cpp:
```diff
@@ -349,6 +352,35 @@ void MultithreadingRenderPasses::record_image_memory_barriers(vkb::CommandBuffer
+	memory_barrier.new_layout      = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+	memory_barrier.src_access_mask = 0;
+	memory_barrier.dst_access_mask = VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_READ_BIT | VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT;
+	memory_barrier.src_stage_mask  = VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT;
```
should this also be a fragment stage? @cforfang

ChristianF-ARM replied:
Looks OK; it’s a WAR hazard (last use of the image was a read some frames ago), so no memory dependency needed, and the execution dependency is satisfied through the fence wait before the images for this frame-index is re-used.

PavelRudko-ARM committed:
Removed pointing to better performance with secondary buffers

gary-sweet commented on samples/performance/multithreading_render_passes/multithreading_render_passes.cpp:
```diff
@@ -52,7 +52,7 @@ bool MultithreadingRenderPasses::prepare(vkb::Platform &platform)
 	}
 
-	load_scene("scenes/sponza/Sponza01.gltf");
+	load_scene("scenes/bonza/Bonza4x.gltf");
```
This is still wrong I'm afraid. It should be Bonza4X not Bonza4x (upper vs lower case x)

PavelRudko-ARM replied:
Fixed :)

PavelRudko-ARM committed:
Fixed scene path

PavelRudko-ARM committed:
Fixed building after rebase

r-potter commented:
![title](https://user-images.githubusercontent.com/6206191/86583128-a72ed100-bf7a-11ea-91d5-d9ba80e064ac.png)

It would be good to do something with the title here. Either add some line breaks or shorten it.

r-potter commented on samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md:
```diff
@@ -0,0 +1,99 @@
+## Profiling
+
+Android Profiler can help to see how threads are utilized. Flame Chart shows how much time was spent for each function execution during a particular timeframe. In this particular example total contribution of command buffers recording in the main thread is 9.94 seconds within a 10 seconds capture with multi-threading disabled. 
+
```
Can we make this a bit more generic and less android specific? i.e. "A profiling tool, such as Android Profiler, can help..."

r-potter commented on samples/performance/multithreading_render_passes/multithreading_render_passes_tutorial.md:
```diff
@@ -0,0 +1,99 @@
+That means, two threads perform the same amount of work in 10s as one thread in more than 15.7 seconds and we should expect approximately 1.57 times better performance. 
+
+And indeed, in debug build, which was used for profiling, frame time is decreased from 531.1ms to 337.7ms using multi-threading (1.57 times decrease).
+
```
If part of the purpose of the samples is to illustrate best practice, we probably shouldn't be using debug builds for profiling.

PavelRudko-ARM replied:
The purpose of using a debuggable package for profiling was to see function execution times in the Flame Chart. The frame times in the section above, however, were measured in a release build.

PavelRudko-ARM committed:
Tutorial update (profiling tools)

JoseEmilio-ARM commented:
> ![title](https://user-images.githubusercontent.com/6206191/86583128-a72ed100-bf7a-11ea-91d5-d9ba80e064ac.png)
> 
> It would be good to do something with the title here. Either add some line breaks or shorten it.

Shorter term we could remove the 'GPU: ' prefix from the GPU name label to gain a bit more space, and in this case we could perhaps shorten the sample name to 'Multithreading RP'

However this will also be the case for future and present long name samples like 'Specialization constants', so maybe it would be better to raise an issue to fix it in a different PR? Some options like the ones you mentioned are certainly possible but require a bit of fiddling with Imgui to programatically adjust the width and height of the sample name given the width of the screen and the width of the GPU label.

r-potter reviewed (changes requested):
UI/title still overlaps. Shortening the title/GPU name label would help
UI is very sluggish and occasionally doesn't respond due to the long frame durations (350+ms on some devices)
Sample crashes on launch on S7

PavelRudko-ARM committed:
Rebase master

PavelRudko-ARM committed:
Fragment shader and subpasses fix after merge

PavelRudko-ARM committed:
Shader include fix

PavelRudko-ARM committed:
Shorter title

PavelRudko-ARM commented:
The title now should not overlap other UI elements.
![short_title](https://user-images.githubusercontent.com/60350359/90119995-2f5e7e00-dd52-11ea-8a52-ac4b056afbf0.png)

carlosadcq commented:
Tested on Qualcomm platforms. Built properly and ran as expected. I can confirm titles don't overlap anymore.

r-potter commented:
I was asked to check that I was testing a release build of this due to the sluggish UI performance. 

Release builds appear to trigger an error with the resource caching system:

````
[info] Loaded gltf image #15 (vase_hanging.ktx)
[info] Loaded gltf image #24 (sponza_roof_diff.ktx)
[info] Time spent loading images: 0.043769 seconds across 8 threads.
[error] [ching.h:719] Creation error for #1 cache object (N3vkb12ShaderModuleE)
[error] ['n' exceeds maximum supported size:75] std::exception
[info] ContentRectChanged: 0x7c5a58da80

[info] Closing App (Runtime: 0.7)
[info] Total device memory leaked: 0 bytes.
````

This occurs on both Qualcomm and Mali devices.

PavelRudko-ARM committed:
Merge branch 'master' of https://github.com/KhronosGroup/Vulkan-Samples into multithreading_render_passes_release

PavelRudko-ARM commented:
> I was asked to check that I was testing a release build of this due to the sluggish UI performance.
> 
> Release builds appear to trigger an error with the resource caching system:
> 
> ```
> [info] Loaded gltf image #15 (vase_hanging.ktx)
> [info] Loaded gltf image #24 (sponza_roof_diff.ktx)
> [info] Time spent loading images: 0.043769 seconds across 8 threads.
> [error] [ching.h:719] Creation error for #1 cache object (N3vkb12ShaderModuleE)
> [error] ['n' exceeds maximum supported size:75] std::exception
> [info] ContentRectChanged: 0x7c5a58da80
> 
> [info] Closing App (Runtime: 0.7)
> [info] Total device memory leaked: 0 bytes.
> ```
> 
> This occurs on both Qualcomm and Mali devices.

I have just rebased to the latest master and tried running it on Huawei P20 with validation layers on. Everything works fine. The error for #1 cache object means an empty fragment shader is not getting compiled. Could you please tell an example of a device? Thanks!

r-potter commented:
> I have just rebased to the latest master and tried running it on Huawei P20 with validation layers on. Everything works fine. The error for #1 cache object means an empty fragment shader is not getting compiled. Could you please tell an example of a device? Thanks!

I saw it on both an Adreno 630 (Note 9) and a Mali-based S10

PavelRudko-ARM commented:
> > I have just rebased to the latest master and tried running it on Huawei P20 with validation layers on. Everything works fine. The error for #1 cache object means an empty fragment shader is not getting compiled. Could you please tell an example of a device? Thanks!
> 
> I saw it on both an Adreno 630 (Note 9) and a Mali-based S10

I have just tested it on an S10 with Mali and the sample works fine. Could you, please, make sure you test it with up-to-date sources and assets? I have rebased to latest master, so it should be synced now.

r-potter commented:
A completely clean checkout appears to work fine. The sluggish performance is indeed only on debug builds. Release builds give a more reasonable 25-40ms per frame. I'm happy with this MR now.
